### ISSUE TITLE
Revisit clang-format

### ISSUE THREAD

**Issue by maruncz** [darktable-org/darktable#11397]
We discussed in #11387, if we enclose all problematic snippets if `// clang-format off` and `// clang-format on`, we can revisit clang-formatting.

My firts suggestion is to reduce ColumnLimit to 80-90.

**Comment by TurboGit** [comment]
I'd like to see the style for:
- routine declaration
- some part of code of before / after

This will help stating the discussion. TIA.

**Comment by maruncz** [comment]
Parameter reference here: https://clang.llvm.org/docs/ClangFormatStyleOptions.html
routine declaration is modified by: `BinPackArguments` and `AlignAfterOpenBracket` both has demonstrations in reference

**Comment by maruncz** [comment]
parameter reference: https://clang.llvm.org/docs/ClangFormatStyleOptions.html
Routine declaration is modified by `AllowAllParametersOfDeclarationOnNextLine` and `BinPackParameters`
Current settings are both true. This produces declaration like this: 
```
static int show_pango_text(dt_bauhaus_widget_t *w, GtkStyleContext *context,
                           cairo_t *cr, const char *text, float x_pos,
                           float y_pos, float max_width, gboolean right_aligned,
                           gboolean calc_only, PangoEllipsizeMode ellipsize,
                           gboolean is_markup, gboolean is_label, float *width,
                           float *height)
```

if we set `BinPackParameters` to false:
```
static int show_pango_text(dt_bauhaus_widget_t *w,
                           GtkStyleContext *context,
                           cairo_t *cr,
                           const char *text,
                           float x_pos,
                           float y_pos,
                           float max_width,
                           gboolean right_aligned,
                           gboolean calc_only,
                           PangoEllipsizeMode ellipsize,
                           gboolean is_markup,
                           gboolean is_label,
                           float *width,
                           float *height)
```

if i understand reference correctly, `AllowAllParametersOfDeclarationOnNextLine=true` allow parameters to go on next line to fit column limit. If parameters are too long they will get split into multiple lines anyway.
when is `BinPackParameters=false` and parameters wont fit column limit, it will force every parameter on single line, othervise there can be multiple parameters per line.

**Comment by TurboGit** [comment]
Do you think we can have:

```
static int show_pango_text
   (dt_bauhaus_widget_t *w,
    GtkStyleContext *context,
    cairo_t *cr,
    const char *text,
    float x_pos,
    float y_pos,
    float max_width,
    gboolean right_aligned,
    gboolean calc_only,
    PangoEllipsizeMode ellipsize,
    gboolean is_markup,
    gboolean is_label,
    float *width,
    float *height)
```

I think it would be better as in dt we have some very long function name and so it will be moving the list of parameters very far of the right (probably also defeating the 80 chars line limit).

**Comment by maruncz** [comment]
with `BinPackParameters: false` and `AllowAllParametersOfDeclarationOnNextLine: false` i'm able to do this
```
static void super_long_function_name_fsghfdghgfdhffg(dt_bauhaus_widget_t *w,
                                                     GtkStyleContext *context,
                                                     cairo_t *cr)
{
}

static void super_long_function_name_alkjfdshlksajh(dt_bauhaus_widget_t *w,
                                                    GtkStyleContext *context,
                                                    cairo_t *cr,
                                                    const char *text,
                                                    float x_pos,
                                                    float y_pos,
                                                    float max_width,
                                                    gboolean right_aligned,
                                                    gboolean calc_only,
                                                    PangoEllipsizeMode ellipsize,
                                                    gboolean is_markup,
                                                    gboolean is_label,
                                                    float *width,
                                                    float *height)
{
}

static void super_long_function_name_kdjhgflkjdshgflkjhkjhdlkjhgflkjdshgflkjhk(
    dt_bauhaus_widget_t *w,
    GtkStyleContext *context,
    cairo_t *cr,
    const char *text,
    float x_pos,
    float y_pos,
    float max_width,
    gboolean right_aligned,
    gboolean calc_only,
    PangoEllipsizeMode ellipsize,
    gboolean is_markup,
    gboolean is_label,
    float *width,
    float *height)
{
}
```

**Comment by maruncz** [comment]
effect on shorter functions:
```
static void function_name(dt_bauhaus_widget_t *w) {}

static void
function_name(dt_bauhaus_widget_t *w, GtkStyleContext *context, cairo_t *cr)
{
}

static void super_long_function_name(dt_bauhaus_widget_t *w,
                                     GtkStyleContext *context,
                                     cairo_t *cr)
{
}
```

**Comment by TurboGit** [comment]
The effect on shorter functions is a bit unfortunate, I would have preferred to have a uniform way of indenting/styling.

**Comment by maruncz** [comment]
I don't know how to make it consistent

**Comment by TurboGit** [comment]
> I don't know how to make it consistent

No big deal :) I can live with that.

**Comment by TurboGit** [comment]
Can you post here some snippet of code that you feel really wrong in master code-base and now reformatted by the tools? This will help the discussion moving forward. TIA.

**Comment by maruncz** [comment]
I can't find it, but there are places wit declaration of multiple variables at the same line even with initialization, which i don't think are readable.

and there is lot of small things, like `AlignConsecutiveAssignments`, `SpaceBeforeParens`,

**Comment by AlicVB** [comment]
Just to know, what is the final goal here ?
I ask that because I'm not a big fan of a pass of such script on the whole code. That has been done in the past, and that make git history very very hard to follow and track down. For example, if you try to use git blame for quite old code part, you are quite sure to have a "code formating" commit. and considering the amount of change in such commits, it's sometimes very hard to go more in the past...

That said, I use for quite some times now the "hook" solution, and I find that quite useful, although I would prefer to have more finetune rules, like what is done here...

**Comment by maruncz** [comment]
It originally started because some statements are so long i have to scroll vertically to read them and current rules wouldn't solve this (long column limit).
my goal is to have clang format that i can use on my code.
in my projects i used to format whole file at the same time. but QtCreator has feature to only format selection, and i used that in rawspeed.
I'm also not fan of format commits.
we can use `clang-format-diff`: https://clang.llvm.org/docs/ClangFormat.html#script-for-patch-reformatting in hook so only changed code would get formatted

**Comment by AlicVB** [comment]
> we can use `clang-format-diff`: https://clang.llvm.org/docs/ClangFormat.html#script-for-patch-reformatting in hook so only changed code would get formatted

that's what I'm doing actually (with very basics rules) but I'm not sure it's possible to implement that so devs use that without setting it themself...

**Comment by maruncz** [comment]
after quick search i think that hooks cannot be set by repository due to safety reasons

**Comment by maruncz** [comment]
Hi is there any update from your side?

I think we agree that formatting commits are bad.

I looked around in darktable and i found mention of clang-format and git hooks in contributing file.
This file though is not mentioned in readme, instead there is chapter with link to darktable website and wiki, but none of them mention clang-format and on website there is link to wiki, but different.
I think this should be updated and made consistent.

I think that we should change formatting gradually, i personally reformat only code i work in.
We could change formatting script so i runs `clang-format-diff` and add mention somewhere.
If someone submits PR we could ask if he/she did format and if not, he/she could make clang-format-diff between begining and end of PR.

**Comment by TurboGit** [comment]
> If someone submits PR we could ask if he/she did format and if not, he/she could make clang-format-diff between begining and end of PR.

No because many people don't know how to split commits for the same file and this will make hard or impossible to review the diff. So, I do prefer reformatting the files in batch, all of them first. And then ensure that we keep good style while working and if not I could run the script on the new files to "fix" the style.

**Comment by maruncz** [comment]
I don't have strong opinion on any solution so do as you like, but i would like to have this issure resolved, either by making new clang format and using it, or not.
